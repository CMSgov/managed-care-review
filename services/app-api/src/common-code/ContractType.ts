import type {
    Contract,
    ContractPackageSubmission,
    ContractRevision,
} from '../gen/gqlServer'

const getLastContractSubmission = (
    contract: Contract
): ContractPackageSubmission | undefined => {
    return (
        (contract.packageSubmissions && contract.packageSubmissions[0]) ??
        undefined
    )
}

const getContractRev = (contract: Contract): ContractRevision | undefined => {
    if (contract.draftRevision) {
        return contract.draftRevision
    } else {
        return getLastContractSubmission(contract)?.contractRevision
    }
}
const isContractOnly = (contract: Contract): boolean => {
    const contractRev = getContractRev(contract)
    return contractRev?.formData?.submissionType === 'CONTRACT_ONLY'
}

const isBaseContract = (contract: Contract): boolean => {
    const contractRev = getContractRev(contract)
    return contractRev?.formData?.contractType === 'BASE'
}

const isContractAmendment = (contract: Contract): boolean => {
    const contractRev = getContractRev(contract)
    return contractRev?.formData?.contractType === 'AMENDMENT'
}

const isCHIPOnly = (contract: Contract): boolean => {
    const contractRev = getContractRev(contract)
    return contractRev?.formData?.populationCovered === 'CHIP'
}

const isContractAndRates = (contract: Contract): boolean => {
    const contractRev = getContractRev(contract)
    return contractRev?.formData?.submissionType === 'CONTRACT_AND_RATES'
}

const isContractWithProvisions = (contract: Contract): boolean =>
    isContractAmendment(contract) ||
    (isBaseContract(contract) && !isCHIPOnly(contract))

const isSubmitted = (contract: Contract): boolean =>
    contract.status === 'SUBMITTED'

export {
    isContractWithProvisions,
    isBaseContract,
    isContractAmendment,
    isCHIPOnly,
    isContractOnly,
    isContractAndRates,
    isSubmitted,
}
