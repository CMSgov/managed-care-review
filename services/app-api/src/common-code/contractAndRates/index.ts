export {
    isBaseContract,
    isCHIPOnly,
    isContractOnly,
    isContractAndRates,
    isContractAmendment,
    isSubmitted,
    isContractWithProvisions,
    getLastContractSubmission,
} from './contractData'

export {
    getProvisionDictionary,
    sortModifiedProvisions,
    generateApplicableProvisionsList,
    generateProvisionLabel,
    isMissingProvisions,
    hasValidModifiedProvisions,
} from './ContractProvisions'

export {
    modifiedProvisionMedicaidBaseKeys,
    modifiedProvisionMedicaidAmendmentKeys,
    provisionCHIPKeys,
    isCHIPProvision,
    isMedicaidAmendmentProvision,
    isMedicaidBaseProvision,
} from './ModifiedProvisions'
