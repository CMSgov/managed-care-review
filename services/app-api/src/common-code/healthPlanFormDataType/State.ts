type StateType = {
    code: string
    name: string
    programs: ProgramArgType[]
}

interface ProgramArgType {
    id: string
    name: string // short name. This is used most often in the application including in submission name
    fullName: string // full name is used in submission summary page
    isRateProgram: boolean // specifies if program relates to rates rather than contract
}

export type { StateType, ProgramArgType }
