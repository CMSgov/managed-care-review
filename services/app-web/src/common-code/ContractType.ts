import { Contract, ContractRevision, UnlockedContract } from '../gen/gqlClient'
import { getLastContractSubmission } from '../gqlHelpers/contractsAndRates'

const getContractRev = (contract: Contract | UnlockedContract): ContractRevision | undefined => {
    if (contract.draftRevision) {
        return contract.draftRevision
    } else {
        return getLastContractSubmission(contract)?.contractRevision
    }
}
const isContractOnly = (contract: Contract | UnlockedContract): boolean => {
    const contractRev = getContractRev(contract)
    return contractRev?.formData?.submissionType === 'CONTRACT_ONLY' 
}


const isBaseContract = (contract: Contract | UnlockedContract): boolean => {
    const contractRev = getContractRev(contract)
    return contractRev?.formData?.contractType === 'BASE'
}

const isContractAmendment = (contract: Contract | UnlockedContract): boolean => {
    const contractRev = getContractRev(contract)
    return contractRev?.formData?.contractType === 'AMENDMENT'
}

const isCHIPOnly = (contract: Contract | UnlockedContract): boolean => {
    const contractRev = getContractRev(contract)
    return contractRev?.formData?.populationCovered === 'CHIP'
}

const isContractAndRates = (contract: Contract | UnlockedContract): boolean => {
    const contractRev = getContractRev(contract)
    return contractRev?.formData?.submissionType === 'CONTRACT_AND_RATES'
}

const isContractWithProvisions = (contract: Contract | UnlockedContract): boolean =>
    isContractAmendment(contract) || (isBaseContract(contract) && !isCHIPOnly(contract))

const isSubmitted = (contract: Contract | UnlockedContract): boolean =>
    contract.status === 'SUBMITTED'

export {
    isContractWithProvisions,
    isBaseContract,
    isContractAmendment,
    isCHIPOnly,
    isContractOnly,
    isContractAndRates,
    isSubmitted,
}
