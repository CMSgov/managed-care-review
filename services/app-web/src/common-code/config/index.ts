export { assertIsAuthMode, assertNever } from './config'
export type { AuthModeType } from './config'
