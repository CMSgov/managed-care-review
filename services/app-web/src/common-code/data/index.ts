export {removeNullAndUndefined, replaceNullsWithUndefineds } from './dataUtilities'