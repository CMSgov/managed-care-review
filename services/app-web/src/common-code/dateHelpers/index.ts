export { formatCalendarDate, formatRateNameDate } from './calendarDate'
export { formatGQLDate } from './gqlDate'
export { dayjs } from './dayjs'
