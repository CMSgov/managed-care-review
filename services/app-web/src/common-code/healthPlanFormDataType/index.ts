// TEMPORARY: these files are embedded in app-web for now b/c
// CRA prevents you from importing code outside of /src
// The fix is to use yarn workspaces to allow us to import shared packages
// Domain Types
export type {
    ActuarialFirmType,
    ActuaryCommunicationType,
    ActuaryContact,
    ContractAmendmentInfo,
    ContractExecutionStatus,
    ContractType,
    DocumentCategoryType,
    UnlockedHealthPlanFormDataType,
    ManagedCareEntity,
    RateType,
    RateCapitationType,
    StateContact,
    SubmissionDocument,
    SubmissionType,
    RateInfoType,
    PopulationCoveredType,
    SharedRateCertDisplay,
} from './UnlockedHealthPlanFormDataType'

export type {
    FederalAuthority,
    CHIPFederalAuthority,
} from './FederalAuthorities'

export type {
    ModifiedProvisionsMedicaidAmendment,
    ModifiedProvisionsMedicaidBase,
    CHIPModifiedProvisions,
    GeneralizedProvisionType,
    MedicaidAmendmentProvisionType,
    GeneralizedModifiedProvisions,
} from './ModifiedProvisions'

export {
    federalAuthorityKeys,
    federalAuthorityKeysForCHIP,
} from './FederalAuthorities'

export {
    modifiedProvisionMedicaidBaseKeys,
    modifiedProvisionMedicaidAmendmentKeys,
    provisionCHIPKeys,
    isCHIPProvision,
    isMedicaidAmendmentProvision,
} from './ModifiedProvisions'

export type { LockedHealthPlanFormDataType } from './LockedHealthPlanFormDataType'
export {
    hasValidContract,
    hasValidDocuments,
    hasValidRates,
    hasAnyValidRateData,
    hasValidSupportingDocumentCategories,
    isContractAndRates,
    isContractOnly,
    isUnlockedHealthPlanFormData,
    isLockedHealthPlanFormData,
    isSubmitted,
    programNames,
    packageName,
    generateRateName,
    convertRateSupportingDocs,
    removeInvalidProvisionsAndAuthorities,
    removeRatesData,
    hasValidPopulationCoverage,
    isValidAndCurrentLockedHealthPlanFormData,
} from './healthPlanFormData'

export type { HealthPlanFormDataType } from './HealthPlanFormDataType'
export type { ProgramArgType, StateType } from './State'

export { StateCodes, isValidStateCode } from './StateCodeType'
export type { StateCodeType } from './StateCodeType'
