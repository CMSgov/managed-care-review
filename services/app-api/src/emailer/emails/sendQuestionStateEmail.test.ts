import {
    testEmailConfig,
    mockContractRev,
    mockMNState,
} from '../../testHelpers/emailerHelpers'
import type {
    CMSUserType,
    ContractRevisionWithRatesType,
    StateType,
} from '../../domain-models'
import type { ContractFormDataType } from '../../domain-models'
import { packageName } from 'app-web/src/common-code/healthPlanFormDataType'
import { sendQuestionStateEmail } from './index'

const defaultSubmitters = ['submitter1@example.com', 'submitter2@example.com']

const flState: StateType = {
    stateCode: 'FL',
    name: 'Florida',
}

const cmsUser: CMSUserType = {
    id: '1234',
    role: 'CMS_USER',
    divisionAssignment: 'DMCO',
    familyName: 'McDonald',
    givenName: 'Ronald',
    email: 'cms@email.com',
    stateAssignments: [flState],
}

const formData: ContractFormDataType = {
    programIDs: ['abbdf9b0-c49e-4c4c-bb6f-040cb7b51cce'],
    populationCovered: 'CHIP',
    submissionType: 'CONTRACT_AND_RATES',
    riskBasedContract: false,
    submissionDescription: 'A submitted submission',
    supportingDocuments: [
        {
            s3URL: 'bar',
            name: 'foo',
            sha256: 'fakesha',
        },
    ],
    contractType: 'BASE',
    contractExecutionStatus: undefined,
    contractDocuments: [
        {
            s3URL: 'bar',
            name: 'foo',
            sha256: 'fakesha',
        },
    ],
    contractDateStart: new Date('01/01/2024'),
    contractDateEnd: new Date('01/01/2025'),
    managedCareEntities: ['MCO'],
    federalAuthorities: ['VOLUNTARY', 'BENCHMARK'],
    inLieuServicesAndSettings: undefined,
    modifiedBenefitsProvided: undefined,
    modifiedGeoAreaServed: undefined,
    modifiedMedicaidBeneficiaries: undefined,
    modifiedRiskSharingStrategy: undefined,
    modifiedIncentiveArrangements: undefined,
    modifiedWitholdAgreements: undefined,
    modifiedStateDirectedPayments: undefined,
    modifiedPassThroughPayments: undefined,
    modifiedPaymentsForMentalDiseaseInstitutions: undefined,
    modifiedMedicalLossRatioStandards: undefined,
    modifiedOtherFinancialPaymentIncentive: undefined,
    modifiedEnrollmentProcess: undefined,
    modifiedGrevienceAndAppeal: undefined,
    modifiedNetworkAdequacyStandards: undefined,
    modifiedLengthOfContract: undefined,
    modifiedNonRiskPaymentArrangements: undefined,
    statutoryRegulatoryAttestation: undefined,
    statutoryRegulatoryAttestationDescription: undefined,
    stateContacts: [
        {
            name: 'test1',
            titleRole: 'Foo1',
            email: 'test1@example.com',
        },
        {
            name: 'test2',
            titleRole: 'Foo2',
            email: 'test2@example.com',
        },
    ],
}
const dateAsked = new Date('01/01/2024')
test('to addresses list includes submitter emails', async () => {
    const sub = mockContractRev()
    const defaultStatePrograms = mockMNState().programs
    const template = await sendQuestionStateEmail(
        sub,
        defaultSubmitters,
        cmsUser,
        testEmailConfig(),
        defaultStatePrograms,
        dateAsked
    )

    if (template instanceof Error) {
        throw template
    }

    expect(template).toEqual(
        expect.objectContaining({
            toAddresses: expect.arrayContaining(defaultSubmitters),
        })
    )
})

test('to addresses list includes all state contacts on submission', async () => {
    const sub: ContractRevisionWithRatesType = {
        ...mockContractRev({ formData }),
    }
    const defaultStatePrograms = mockMNState().programs
    const template = await sendQuestionStateEmail(
        sub,
        defaultSubmitters,
        cmsUser,
        testEmailConfig(),
        defaultStatePrograms,
        dateAsked
    )

    if (template instanceof Error) {
        console.error(template)
        return
    }

    sub.formData.stateContacts.forEach((contact) => {
        expect(template).toEqual(
            expect.objectContaining({
                toAddresses: expect.arrayContaining([contact.email]),
            })
        )
    })
})

test('to addresses list does not include duplicate state receiver emails on submission', async () => {
    const formDataWithDuplicateStateContacts = {
        ...formData,
        stateContacts: [
            {
                name: 'test1',
                titleRole: 'Foo1',
                email: 'test1@example.com',
            },
            {
                name: 'test1',
                titleRole: 'Foo1',
                email: 'test1@example.com',
            },
        ],
    }
    const sub: ContractRevisionWithRatesType = {
        ...mockContractRev({ formData: formDataWithDuplicateStateContacts }),
    }
    const defaultStatePrograms = mockMNState().programs
    const template = await sendQuestionStateEmail(
        sub,
        defaultSubmitters,
        cmsUser,
        testEmailConfig(),
        defaultStatePrograms,
        dateAsked
    )

    if (template instanceof Error) {
        console.error(template)
        return
    }

    expect(template.toAddresses).toEqual([
        'test1@example.com',
        ...defaultSubmitters,
        ...testEmailConfig().devReviewTeamEmails,
    ])
})

test('subject line is correct and clearly states submission is complete', async () => {
    const sub = mockContractRev()
    const defaultStatePrograms = mockMNState().programs
    const name = packageName(
        sub.contract.stateCode,
        sub.contract.stateNumber,
        sub.formData.programIDs,
        defaultStatePrograms
    )

    const template = await sendQuestionStateEmail(
        sub,
        defaultSubmitters,
        cmsUser,
        testEmailConfig(),
        defaultStatePrograms,
        dateAsked
    )

    if (template instanceof Error) {
        console.error(template)
        return
    }

    expect(template).toEqual(
        expect.objectContaining({
            subject: expect.stringContaining(`New questions about ${name}`),
            bodyText: expect.stringContaining(`${name}`),
        })
    )
})

test('includes link to submission', async () => {
    const sub = mockContractRev()
    const defaultStatePrograms = mockMNState().programs
    const template = await sendQuestionStateEmail(
        sub,
        defaultSubmitters,
        cmsUser,
        testEmailConfig(),
        defaultStatePrograms,
        dateAsked
    )

    if (template instanceof Error) {
        console.error(template)
        return
    }

    expect(template).toEqual(
        expect.objectContaining({
            bodyText: expect.stringContaining(
                'Open the submission in MC-Review to answer questions'
            ),
            bodyHTML: expect.stringContaining(
                `href="http://localhost/submissions/${sub.contract.id}"`
            ),
        })
    )
})

test('includes information about what to do next', async () => {
    const sub = mockContractRev()
    const defaultStatePrograms = mockMNState().programs
    const template = await sendQuestionStateEmail(
        sub,
        defaultSubmitters,
        cmsUser,
        testEmailConfig(),
        defaultStatePrograms,
        dateAsked
    )

    if (template instanceof Error) {
        console.error(template)
        return
    }

    expect(template).toEqual(
        expect.objectContaining({
            bodyText: expect.stringContaining(
                'You must answer the question before CMS can continue reviewing it'
            ),
        })
    )
})

test('includes expected data on the CMS analyst who sent the question', async () => {
    const sub = mockContractRev()
    const defaultStatePrograms = mockMNState().programs

    const template = await sendQuestionStateEmail(
        sub,
        defaultSubmitters,
        cmsUser,
        testEmailConfig(),
        defaultStatePrograms,
        dateAsked
    )

    if (template instanceof Error) {
        console.error(template)
        return
    }

    expect(template).toEqual(
        expect.objectContaining({
            bodyText: expect.stringContaining(
                'Sent by: Ronald McDonald (DMCO)  cms@email.com (cms@email.com)'
            ),
        })
    )
    expect(template).toEqual(
        expect.objectContaining({
            bodyText: expect.stringContaining('Date: 01/01/2024'),
        })
    )
})

test('renders overall email for a new question as expected', async () => {
    const sub = mockContractRev()
    const defaultStatePrograms = mockMNState().programs
    const result = await sendQuestionStateEmail(
        sub,
        defaultSubmitters,
        cmsUser,
        testEmailConfig(),
        defaultStatePrograms,
        dateAsked
    )

    if (result instanceof Error) {
        console.error(result)
        return
    }

    expect(result.bodyHTML).toMatchSnapshot()
})
