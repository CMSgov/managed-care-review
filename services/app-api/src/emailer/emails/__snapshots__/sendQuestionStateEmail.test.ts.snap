// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders overall email for a new question as expected 1`] = `
"CMS asked questions about MCR-MN-0003-SNBC<br />
<b>Sent by:</b> Ronald McDonald (DMCO) <a href="cms@email.com">cms@email.com</a>
<br />
<b>Date:</b> 01/01/2024<br />
<br />
You must respond to the questions before CMS can continue their review.<br />
<br />
<a href="http://localhost/submissions/12345/question-and-answers">Open the submission in MC-Review to answer questions</a>

"
`;
